includes/class-amp-reader-themes.php

```diff
@@ -0,0 +1,313 @@
+<?php
+/**
+ * Fetches and formats data for AMP reader themes.
+ *
+ * @package AMP
+ * @since 1.6.0
+ */
+
+/**
+ * Class AMP_Reader_Themes.
+ *
+ * @since 1.6.0
+ */
+final class AMP_Reader_Themes {
+	/**
+	 * Formatted theme data.
+	 *
+	 * @var array
+	 */
+	private $themes;
+
+	/**
+	 * Reader themes supported by default.
+	 *
+	 * @var array
+	 */
+	private $default_reader_themes;
+
+	/**
+	 * Whether themes can be installed in the current WordPress installation.
+	 *
+	 * @var bool
+	 */
+	private $can_install_themes;
+
+	/**
+	 * The default reader theme.
+	 *
+	 * @var string
+	 */
+	const DEFAULT_READER_THEME = 'classic';
+
+	/**
+	 * Status indicating a reader theme is active on the site.
+	 *
+	 * @var string
+	 */
+	const STATUS_ACTIVE = 'active';
+
+	/**
+	 * Status indicating a reader theme is installed but not active.
+	 *
+	 * @var string
+	 */
+	const STATUS_INSTALLED = 'installed';
+
+	/**
+	 * Status indicating a reader theme is not installed but is installable.
+	 *
+	 * @var string
+	 */
+	const STATUS_INSTALLABLE = 'installable';
+
+	/**
+	 * Status indicating a reader theme is not installed and can't be installed.
+	 *
+	 * @var string
+	 */
+	const STATUS_NON_INSTALLABLE = 'non-installable';
+
+	/**
+	 * Retrieves all AMP plugin options specified in the endpoint schema.
+	 *
+	 * @return array Formatted theme data.
+	 */
+	public function get_themes() {
+		if ( null !== $this->themes ) {
+			return $this->themes;
+		}
+
+		$themes = $this->get_default_reader_themes();
+
+		/**
+		 * Filters supported reader themes.
+		 *
+		 * @param array $themes [
+		 *     Reader theme data.
+		 *     {
+		 *         @type string         $name           Theme name.
+		 *         @type string         $slug           Theme slug.
+		 *         @type string         $slug           URL of theme preview.
+		 *         @type string         $screenshot_url The URL of a mobile screenshot. Note: if this is empty, the theme may not display.
+		 *         @type string         $homepage       A link to a page with more information about the theme.
+		 *         @type string         $description    A description of the theme.
+		 *         @type string|boolean $requires       Minimum version of WordPress required by the theme. False if all versions are supported.
+		 *         @type string|boolean $requires_php   Minimum version of PHP required by the theme. False if all versions are supported.
+		 *         @type string         $download_link  A link to the theme's zip file. If empty, the plugin will attempt to download the theme from wordpress.org.
+		 *     }
+		 * ]
+		 */
+		$themes = array_filter( (array) apply_filters( 'amp_reader_themes', $themes ) );
+
+		$themes = array_filter(
+			$themes,
+			function( $theme ) {
+				return is_array( $theme ) && ! empty( $theme ) && ! empty( $theme['screenshot_url'] ); // Screenshots are required.
+			}
+		);
+
+		$themes = array_map(
+			function ( $theme ) {
```

Comment:
```suggestion
			static function ( $theme ) {
```

Reply:
I believe `$this` inside the callback would break in this case, correct?

Reply:
Ah yes, your right :smile:.